### DISCUSSION TITLE
curl_multi_perform crashing when run on a system with OS Hardening.

### DISCUSSION THREAD

**Discussion by bansalnehal** [curl/curl#13259] [Q&A]
Hi, I am using curl version 7.76.0 on my system to send and receive HTTP messages from a server. It works fine when continuous messages at the rate of 100 are being sent and received for a duration of 2-3 days (longest I have observed). However, when OS Hardening is applied on the system, after 15 minutes or so, it is crashing in curl_multi_perform. Could you point out the issue?

**Comment by jay** [comment]
To rule out a bug in curl you should use the latest version. However, since the bug only occurs when OS hardening is applied I'd guess that has something to do with it.

**Reply by bansalnehal** [reply]
Is there a specific area I should check to find out the problem?

**Reply by jay** [reply]
The OS hardening is the specific area. Look at the stack traces.

**Reply by bansalnehal** [reply]
The following is the stack trace. 

/lib64/libpthread.so.0() [0x37c280f7e0]
/usr/lib64/libcrypto.so.10() [0x37c9ce6295]
/usr/lib64/libcrypto.so.10(lh_retrieve+0x1c) [0x37c9ce62ec]
/usr/lib64/libcrypto.so.10() [0x37c9ce8e3e]
/usr/lib64/libcrypto.so.10(ERR_get_state+0x3b) [0x37c9ce912b]
/usr/lib64/libcrypto.so.10(ERR_clear_error+0xd) [0x37c9ce98cd]
/usr/lib64/libcurl.so.4() [0x343c26426a]
/usr/lib64/libcurl.so.4() [0x343c2482bf]
/usr/lib64/libcurl.so.4() [0x343c25b0d0]
/usr/lib64/libcurl.so.4() [0x343c23df2f]
/usr/lib64/libcurl.so.4(curl_multi_perform+0x10a) [0x343c23ec9a]

**Reply by jay** [reply]
This reminds me of your [previous question](https://github.com/curl/curl/discussions/12642#discussioncomment-8626514). Another crash with libcrypto calling into libpthread. We can't do anything to help with this except advise you to build the latest version of OpenSSL and libcurl with debug symbols, follow the [thread safety guidelines](https://curl.se/libcurl/c/threadsafe.html), consider using threadsanitizer

**Comment by dfandrich** [comment]
> Could you point out the issue?o

Not without a whole lot more information. The operating system for one, where
you got your curl for another, and a way to reproduce what you're seeing. What
is "OS Hardening" to start with? What do you mean by "continuous messages"?
What units is "100" in? Can you provide a standalone program that shows the
issue? And even with all that, you're using a version that's 3 years old and
that has had over 3100 bug fixes made to it; there's a good chance whatever it is
you're seeing won't happen on a recent release.

**Reply by bansalnehal** [reply]
Following is the information: 
OS - CentOS 6.10
Continuous messages - There are requests coming on my system at the rate of 100 requests per sec. These requests are made into easy handles, added to a multi handle and sent to the HTTP server using curl_multi_perform. 

Standalone program:
There are two threads:

Thread 1

//Receive requests, make it into an easy handle and add it to a queue. (thread safe)
//Used the following options on the handle
curl_easy_setopt(easy, CURLOPT_HTTPHEADER, headers); \
                 curl_easy_setopt(easy, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_2_0); \
                 curl_easy_setopt(easy, CURLOPT_SSL_VERIFYPEER, 0L); \
                 curl_easy_setopt(easy, CURLOPT_SSL_VERIFYHOST, 0L); \
                 curl_easy_setopt(easy, CURLOPT_URL,url); \
                 curl_easy_setopt(easy, CURLOPT_POSTFIELDSIZE, datalen); \
                 curl_easy_setopt(easy, CURLOPT_COPYPOSTFIELDS, data); \
                 curl_easy_setopt(easy, CURLOPT_WRITEFUNCTION, write_cb); \
                 curl_easy_setopt(easy, CURLOPT_TIMEOUT, 3L);


Thread 2
`
CURLMsg *msg; 
 int msgs_left = -1; 
 int still_alive = 0; 
 do 
 {
       //take one easy handle from easy handle queue and add to multi handle using **curl_multi_add_handle.** (thread-safe)
       int numfds = 0; 
       CURLMcode res = curl_multi_poll(multi ,NULL,0,1000, &numfds);  
       curl_multi_perform(multi, &still_alive); 
       while ((msg = curl_multi_info_read(multi, &msgs_left))) 
       { 
             if (msg->msg == CURLMSG_DONE) 
             { 
                  char *url; 
                  CURL *e = msg->easy_handle; 
                  curl_easy_getinfo(msg->easy_handle, CURLINFO_PRIVATE, &url);                     
                  curl_multi_remove_handle(multi, e); 
                  curl_easy_cleanup(e);

                  //read one handle from easy handle queue and add using **curl_multi_add_handle.**  (thread-safe)
              } 
       }       
} while (1);
`

I am unable to upgrade the curl version due to the dependencies of my OS. Therefore, I would want to fix it in this version.

**Reply by bagder** [reply]
> I am unable to upgrade the curl version due to the dependencies of my OS. Therefore, I would want to fix it in this version.

This is not logical. If you can do any fix at all, you create a fixed version that you apparently can upgrade to/use. That fixed version could then also be an upgraded curl version...

**Comment by dfandrich** [comment]
You've given very little to go on, but "OS Hardening" seems to be important in
some way.

**Comment by dfandrich** [comment]
That's a start of the information we'd need to help. But, it seems that "OS
Hardening" is key here, and you still haven't explained what it is. In any
case, we here wi'll probably just ask you to test the latest version anyway to
check if one of those 3000 bug fixes we've made since your version improved
things, and to verify that one of CentOS's patches to that version haven't
actually caused the problem. If you really want a solution for an old version
and you aren't able to make headway yourself, you're probably going to have to
resort to commercial support.

**Reply by bansalnehal** [reply]
OS hardening here is referring to some security measures for e.g.
1) Unencrypted access methods FTP, telnet disabled on the system.
2) Restricted root access to the system
3) Partitions for /tmp /boot have been separated from /.
